### DISCUSSION TITLE
filter smiles with carboxylic acid functional group

### DISCUSSION THREAD

**Discussion by amir-tagh** [rdkit/rdkit#6620] [Q&A]
Hello,

Is it possible to use rdkit to select a subset of smiles which have carboxylic acid functional group from a parent database.

Thanks,
Amir

**Comment by boglet** [comment]
Iterate through the molecules and use this SMARTS to match COOH:

`[CX3](=O)[OX2H1]`

In code:
`matches = m.GetSubstructMatches(Chem.MolFromSmarts("[CX3](=O)[OX2H1]")`
`if (len(matches)>0):`
`  //do stuff`

**Reply by amir-tagh** [reply]
Thank you.

**Comment by bp-kelley** [comment]
The RDKit has a lot of these built into the FilterCatalog

```
from rdkit import Chem
from rdkit.Chem import FilterCatalog

fc = FilterCatalog.GetFunctionalGroupHierarchy()
m = Chem.MolFromSmiles("CC(=C)C1CCC(=CC1)C(=O)O")

for match in fc.GetMatches(m):
    print(match.GetDescription())
```

Result:
```
CarboxylicAcid
```

To get the whole list of the hierarchy, you can get a python dictionary (name: query molecule) as follows

```
filters = FilterCatalog.GetFlattenedFunctionalGroupHierarchy()
for k, pat in filters.items():
    print(f"{k} -> {Chem.MolToSmarts(pat)}")
```

Results in

```
AcidChloride -> C(=O)Cl
AcidChloride.Aliphatic -> [$(C-&!@[A&!Cl])](=O)Cl
AcidChloride.Aromatic -> [$(C-&!@a)](=O)Cl
Alcohol -> [O&H1&$(O-&!@[#6&!$(C=&!@[O,N,S])])]
Alcohol.Aliphatic -> [O&H1&$(O-&!@[C&!$(C=&!@[O,N,S])])]
Alcohol.Aromatic -> [O&H1&$(O-&!@c)]
Aldehyde -> [C&H1&D2&!$(C-[!#6&!#1])]=O
Aldehyde.Aliphatic -> [C&H1&D2&$(C-&!@C)]=O
Aldehyde.Aromatic -> [C&H1&D2&$(C-&!@a)]=O
Amine -> [N&$(N-[#6])&!$(N-[!#6&!#1])&!$(N-C=[O,N,S])]
Amine.Aliphatic -> [N&!$(N-c)&$(N-C)&!$(N-[!#6&!#1])&!$(N-C=[O,N,S])]
...
```

**Reply by boglet** [reply]
Thanks @bp-kelley - I wasn't aware of this function. I'm sure that it will come in handy for me in the future!
